### PR TITLE
Install the Theia IDE

### PR DESCRIPTION
**Pull request by ishitatsuyuki**
Also did some refactoring to the workspace directory mechanism, so we can just edit an environment variable instead of long sed expression.

Briefly tested.

### CODE REVIEW

File: ubuntu-dev/supervisord.conf [hunk 1]
```diff
@@ -7,7 +7,11 @@ user = user
 command = sudo /usr/sbin/sshd -D
 [program:c9sdk]
 user = user
-command = /home/user/.c9sdk/server.js --auth : --collab --listen 0.0.0.0 --packed --port 8089 --useBrowserCache --workspacetype=janitor -w /home/user
+command = /home/user/.c9sdk/server.js --auth : --collab --listen 0.0.0.0 --packed --port 8089 --useBrowserCache --workspacetype=janitor -w %(ENV_WORKSPACE)s
```
Comment (jankeromnes) [inline ubuntu-dev/supervisord.conf]: Exciting, your `ENV WORKSPACE` variable looks much nicer than our current sed/replace!

But how does `%(ENV_WORKSPACE)` work? (Can we be sure this won't get templated at `COPY` time in `ubuntu-dev.dockerfile`, thus ignore subsequent changes in later Dockerfile instructions like in `thunderbird.dockerfile`?)

File: ubuntu-dev/ubuntu-dev.dockerfile [hunk 2]
```diff
@@ -224,11 +224,20 @@ RUN git clone https://github.com/c9/core.git /home/user/.c9sdk \
 ADD workspace-janitor.js /home/user/.c9sdk/configs/ide/
 RUN sudo chown user:user /home/user/.c9sdk/configs/ide/workspace-janitor.js
 
+# Install the Theia IDE with all features available.
+COPY --chown=user theia /home/user/.theia/
```
Comment (jankeromnes) [inline ubuntu-dev/ubuntu-dev.dockerfile]: Wait, there is a `--chown` option on `COPY`? This is amazing! There are so many places where we `ADD file` or `COPY file` and then `RUN sudo chown user:user file`!

File: ubuntu-dev/supervisord.conf (hunk 1 above)
Comment (ishitatsuyuki) [reply ubuntu-dev/supervisord.conf]: It's a supervisord config syntax, which essentially pulls in the environment variable at runtime.

File: ubuntu-dev/ubuntu-dev.dockerfile (hunk 2 above)
Comment (ishitatsuyuki) [reply ubuntu-dev/ubuntu-dev.dockerfile]: Yes, though it's added recently so keep an eye on your Docker version.

**Comment by jankeromnes** [comment]
Thank you for this! I'd love to help you test this, and I'm very excited about shipping Theia!

Note: It would be nice to `EXPOSE` the Theia port in `ubuntu-dev.dockerfile`, and then add it to all `janitor.json` files out there. (Note: For the future, we thought about making Janitor expose a set of default ports regardless of what each `janitor.json` says, thus allowing project to expose any project-specific ports they like but without having to keep up with `ubuntu-dev`'s default ports and how they change over time).

**Comment by ishitatsuyuki** [comment]
Address the issue. Should be ready to merge.

File: ubuntu-dev/supervisord.conf (hunk 1 above)
Comment (jankeromnes) [reply ubuntu-dev/supervisord.conf]: Ok thanks. Note that `ENV` variables from Dockerfiles don't transpire to runtime environments like Cloud9/VNC terminals. Let's keep an eye out for workspace regressions.

**Pull request merged**